llm/cpp/llm.cpp

```diff
@@ -77,31 +55,17 @@ int main(int argc, char* argv[]) try {
         size_t batch_size = next_tokens.size();
         ireq.get_tensor("input_ids").set_shape({batch_size, 1});
         ov::Tensor attention_mask = ireq.get_tensor("attention_mask");
-        ov::Shape mask_shape = attention_mask.get_shape();
-        mask_shape.at(0) = batch_size;
-        ++mask_shape.at(1);
+        ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(1) + 1};
         attention_mask.set_shape(mask_shape);
         std::fill_n(attention_mask.data<int64_t>(), shape_size(mask_shape), 1);
         ireq.get_tensor("position_ids").set_shape({batch_size, 1});
         std::fill_n(ireq.get_tensor("position_ids").data<int64_t>(), batch_size, mask_shape.at(1) - 1);
-        for (size_t tensor_idx = 3; tensor_idx < inputs.size(); ++tensor_idx) {
-            ov::Shape shape = ireq.get_output_tensor(tensor_idx - 2).get_shape();
-            shape.at(0) = batch_size;
-            ireq.get_input_tensor(tensor_idx).set_shape(shape);
-        }
+        ireq.get_tensor("beam_idx").set_shape({batch_size});
+        int64_t* input_tokens = ireq.get_tensor("input_ids").data<int64_t>();
+        int32_t* beam_idx = ireq.get_tensor("beam_idx").data<int32_t>();
         for (size_t batch_idx = 0; batch_idx < batch_size; ++batch_idx) {
-            ireq.get_tensor("input_ids").data<int64_t>()[batch_idx] = next_tokens.at(batch_idx).token_idx;
-            for (size_t tensor_idx = 3; tensor_idx < inputs.size(); ++tensor_idx) {
-                ov::Tensor present = ireq.get_output_tensor(tensor_idx - 2);
-                ov::Shape present_begin = {next_tokens.at(batch_idx).beam_idx, 0, 0, 0};
-                ov::Shape present_end = present.get_shape();
-                present_end.at(0) = next_tokens.at(batch_idx).beam_idx + 1;
-                ov::Tensor past = ireq.get_input_tensor(tensor_idx);
-                ov::Shape past_begin = {batch_idx, 0, 0, 0};
-                ov::Shape past_end = past.get_shape();
-                past_end.at(0) = batch_idx + 1;
-                ov::Tensor{present, present_begin, present_end}.copy_to(ov::Tensor{past, past_begin, past_end});
-            }
+            input_tokens[batch_idx] = next_tokens.at(batch_idx).token_idx;
+            beam_idx[batch_idx] = next_tokens.at(batch_idx).beam_idx;
```

Comment:
Can we extend API of Tensor to add some sugar: `ireq.get_tensor("beam_idx").copy_from(beam_idx)` where `beam_idx` is `std::vector<int32_t>` in this case. Setting shapes is still required, the number of elements in the tensor and the vector should match.

It would allow us to write something like the following:
```
beam_idx_tensor.copy_from(beams)
input_ids_tensor.copy_from(tokens)
```
if the changes suggested in other comments are applied.

Reply:
We can create more generic ctor which accepts `std::vector`:
```c++
template <typename T>
Tensor(const std::vector<T>& vec); // creates ov::Tensor on pre-allocated memory
```
And then:
```c++
ov::Tensor(beams).copy_to(beam_idx_tensor);
```
How do you think?

Reply:
When you accepting a vector as ctor argument you need to think whether it is copied or shared. The existing ctors in Tensor wrap a memory passed as a pointer. If we do that for a vector it would be error-prone because users will start passing temporary created vector objects right in the ctor because it looks convenient.

If you copy the content instead of sharing, it looks more safe, but it will bring another copy if copy_to is used. We can use set_tensor, but it looks like something that intentionally has more copies underneath, which is probably OK in this particular case. This is not performance critical part in this sample and we can choose any way, but I am afraid that we can introduce bad practices that doesn't scale well with bigger tensors.

Reply:
https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#a23b182c4ffd46abe38f460c1480ae887